lib/handlers/authenticate-handler.ts

```diff
@@ -0,0 +1,273 @@
+    if (!options.model) {
+      throw new InvalidArgumentError('Missing parameter: `model`');
+    }
+
+    if (!options.model.getAccessToken) {
+      throw new InvalidArgumentError(
+        'Invalid argument: model does not implement `getAccessToken()`',
+      );
+    }
+
+    if (options.scope && options.addAcceptedScopesHeader === undefined) {
+      throw new InvalidArgumentError(
+        'Missing parameter: `addAcceptedScopesHeader`',
+      );
+    }
+
+    if (options.scope && options.addAuthorizedScopesHeader === undefined) {
+      throw new InvalidArgumentError(
+        'Missing parameter: `addAuthorizedScopesHeader`',
+      );
+    }
+
+    if (options.scope && !options.model.verifyScope) {
+      throw new InvalidArgumentError(
+        'Invalid argument: model does not implement `verifyScope()`',
+      );
+    }
+
+    this.addAcceptedScopesHeader = options.addAcceptedScopesHeader;
+    this.addAuthorizedScopesHeader = options.addAuthorizedScopesHeader;
+    this.allowBearerTokensInQueryString =
+      options.allowBearerTokensInQueryString;
+    this.model = options.model;
+    this.scope = options.scope;
+  }
+
+  /**
+   * Authenticate Handler.
+   */
+
+  async handle(request: Request, response: Response) {
+    if (!(request instanceof Request)) {
+      throw new InvalidArgumentError(
+        'Invalid argument: `request` must be an instance of Request',
+      );
+    }
+
+    if (!(response instanceof Response)) {
+      throw new InvalidArgumentError(
+        'Invalid argument: `response` must be an instance of Response',
+      );
+    }
+
+    // Extend model object with request
+    this.model.request = request;
+
+    try {
+      let token = await this.getTokenFromRequest(request);
+      token = await this.getAccessToken(token);
+      this.validateAccessToken(token);
+      if (this.scope) {
+        await this.verifyScope(token);
+      }
+      this.updateResponse(response, token);
+
+      return token;
+    } catch (e) {
+      // Include the "WWW-Authenticate" response header field if the client
+      // lacks any authentication information.
+      //
+      // @see https://tools.ietf.org/html/rfc6750#section-3.1
+      if (e instanceof UnauthorizedRequestError) {
+        response.set('WWW-Authenticate', 'Bearer realm="Service"');
+      }
+
+      if (!(e instanceof OAuthError)) {
+        throw new ServerError(e);
+      }
+
+      throw e;
+    }
+  }
+
+  /**
+   * Get the token from the header or body, depending on the request.
+   *
+   * "Clients MUST NOT use more than one method to transmit the token in each request."
+   *
+   * @see https://tools.ietf.org/html/rfc6750#section-2
+   */
+
+  getTokenFromRequest(request: Request) {
+    const headerToken = request.get('Authorization');
+    const queryToken = request.query.access_token;
+    const bodyToken = request.body.access_token;
+
+    if ((!!headerToken && 1) + (!!queryToken && 1) + (!!bodyToken && 1) > 1) {
```

Comment:
This is fairly hard to read, and fails linting:
~~~~
ERROR: ..../project/node-oauth2-server/lib/handlers/authenticate-handler.ts:117:9 - Operands of '+' operation must either be both strings or both numbers or both bigints, but found any + false | 1
~~~~

You can replace it easily with `[headerToken, queryToken, bodyToken].filter(Boolean).length !== 1`.